src/openMVG/features/cudasift/image_describer_cudasift.hpp

```diff
@@ -0,0 +1,157 @@
+#define CSIFT_IMAGE_DESCRIBER_HPP
+
+#include <tuple>
+#include <vector>
+#include "openMVG/image/image.hpp"
+#include "openMVG/features/image_describer.hpp"
+#include "openMVG/features/regions_factory.hpp"
+#include "openMVG/features/feature.hpp"
+#include "third_party/cudasift/cudaImage.h"
+#include "third_party/cudasift/cudaSift.h"
+#include "openMVG/features/cudasift/csift_keypoint.hpp"
+#include <iostream>
+#include <numeric>
+#include <cereal/cereal.hpp>
+#include <Eigen/Core>
+
+namespace openMVG {
+namespace features {
+
+class CSIFT_Image_describer : public Image_describer
+{
+public:
+    struct Params
+    {
+        Params(
+        float initBlur = 1.0f,
+        float thresh = 3.5f,
+		int numOctaves = 5
+        ):
+        initBlur_(initBlur),
+        thresh_(thresh) {}
+
+        template<class Archive>
+        void serialize(Archive & ar)
+        {
+            ar(
+                cereal::make_nvp("initBlur", initBlur_),
+                cereal::make_nvp("thresh", thresh_));
+
+        }
+        float initBlur_;
+        float thresh_;
+		int numOctaves_;
+    };
+
+    CSIFT_Image_describer
+    (
+        const Params params = Params()
+    )
+    :Image_describer(), params_(params)
+    {}
+
+    bool Set_configuration_preset(EDESCRIBER_PRESET preset) override
+    {
+        switch(preset)
+        {
+          case NORMAL_PRESET:
+            params_.thresh_ = 2.5f;
+          break;
+          case HIGH_PRESET:
+            params_.thresh_ = 3.5f;
+          break;
+          case ULTRA_PRESET:
+            params_.thresh_ = 4.00f;
+          break;
+          default:
+            return false;
+        }
+        return true;
+    }
+
+
+    bool Describe(const image::Image<unsigned char>& image,
+    std::unique_ptr<Regions> &regions,
+    const image::Image<unsigned char> * mask = nullptr) override
+    {
+        // Initialize GPU, allocate memory for image
+        InitCuda(0);
+        CudaImage img1;
+        const int w = image.Width(), h = image.Height();
+
+        // Convert to float 
+        const image::Image<float> If(image.GetMat().cast<float>());
+        img1.Allocate(w, h, iAlignUp(w, 128), false, NULL, (float*)If.data());
+        img1.Download();
+        Allocate(regions);
+        SIFT_Regions * regionsCasted = dynamic_cast<SIFT_Regions*>(regions.get());
+        {
+            SiftData siftData1;
+            InitSiftData(siftData1, 32768, true, true);
+			ExtractSift(siftData1, img1, params_.numOctaves_, params_.initBlur_, params_.thresh_, 0.0f, false);
+            int numPts1 = siftData1.numPts;
+			std::cout << "Number of features detected: " << numPts1 << std::endl;
+
+            SiftPoint *sift1 = siftData1.h_data;
+            std::vector< csift::Keypoint > keys;
+            Descriptor<unsigned char, 128> descriptor;
+			// Load keypoint and descriptor data from result, convert descriptor array into Eigen matrix
+            for (int i = 0; i < numPts1 ; i++)
```

Comment:
You can convert directly the keypoint in the right format, you don't need to have this temporary copy for the conversion.